### PR TITLE
Feature/http cache

### PR DESCRIPTION
**Pull request by davidjgonzalez**
Minor adjustments to get it working on my local. It's _awesome_ :)

One change I made was removing the `customAttributes` OSGi Property from HttpCacheConfigImpl. The reasoning is this list of strings may not be sufficient for all customizers. Instead I collapsed the GroupCacheKeyFactory into GroupCacheConfigExtension, and created a impl specific userGroups OSGi Property that both the Extension and CCFactory can leverage.

Technically these 2 classes could be decoupled since they are registered to their respective Service interfaces, and HttpCacheConfig's configs still let you specific the service filter for each individually, but i expect these 2 things (Extension and CCFactory) will be very tightly coupled in most impls.

Other minor changes included handling CacheKey exceptions on key-based invalidation (i have it invalidating the entire cache if that happens)

If you can take a look at the the MemPersistenceObject and make sure the Header logic looks right, would be handy. I was getting an exception when it tried to unwrap the Header values from a wrappedSet to a List.

Made some minor tweaks to the MBean to get it to come up; but think there may be a few issues w it still.

```
25.11.2015 00:07:40.629 *TRACE* [0:0:0:0:0:0:0:1 [1448428060356] GET /content/geometrixx/en.html HTTP/1.1] com.adobe.acs.commons.httpcache.filter.impl.HttpCacheFilter Delivered un-cached request [ /content/geometrixx/en.html ] in 265 ms
25.11.2015 00:07:54.323 *DEBUG* [0:0:0:0:0:0:0:1 [1448428074309] GET /content/geometrixx/en.html HTTP/1.1] com.adobe.acs.commons.httpcache.filter.impl.HttpCacheFilter Delivered cached request [ /content/geometrixx/en.html ] in 13 ms

25.11.2015 00:25:02.799 *TRACE* [0:0:0:0:0:0:0:1 [1448429099625] GET /content/geometrixx/en/events/techsummit.html HTTP/1.1] com.adobe.acs.commons.httpcache.filter.impl.HttpCacheFilter Delivered un-cached request [ /content/geometrixx/en/events/techsummit.html ] in 3171 ms
25.11.2015 00:27:24.362 *DEBUG* [0:0:0:0:0:0:0:1 [1448429244347] GET /content/geometrixx/en/events/techsummit.html HTTP/1.1] com.adobe.acs.commons.httpcache.filter.impl.HttpCacheFilter Delivered cached request [ /content/geometrixx/en/events/techsummit.html ] in 15 ms
```

### CODE REVIEW

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/GroupHttpCacheConfigExtension.java [hunk 1]
```diff
@@ -2,31 +2,54 @@
...
  * at least one of the configured groups is present in the request user's group membership list.
  */
-@Component(immediate = true)
+@Component(immediate = true, metatype = true)
 @Service
-public class GroupHttpCacheConfigExtension implements HttpCacheConfigExtension {
+public class GroupHttpCacheConfigExtension implements HttpCacheConfigExtension, CacheKeyFactory {
     private static final Logger log = LoggerFactory.getLogger(GroupHttpCacheConfigExtension.class);
 
+    // Custom cache config attributes
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/GroupHttpCacheConfigExtension.java:42]: This replaces the customAttributes. It allows custom "extensions" to define whatever set of custom attributes/osgi configs that is necessary.

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/GroupHttpCacheConfigExtension.java [hunk 2]
```diff
@@ -44,14 +70,125 @@ public boolean accepts(SlingHttpServletRequest request, HttpCacheConfig cacheCon
...
                 }
                 return isGroupMatchFound;
             } catch (RepositoryException e) {
-                throw new HttpCacheReposityAccessException("Unable to access group information of request user.", e);
+                throw new HttpCacheRepositoryAccessException("Unable to access group information of request user.", e);
             }
         }
     }
+
+    //-------------------------<CacheKeyFactory methods>
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/GroupHttpCacheConfigExtension.java:88]: Moved GroupCacheKeyFactory into this same class; Allows sharing of the userGroups OSGi Property

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java [hunk 3]
```diff
@@ -111,14 +111,6 @@
                name = "cacheConfigExtension")
     private HttpCacheConfigExtension cacheConfigExtension;
 
-    // Custom cache config attributes
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java:114]: Removed customAttribures as this is the domain of the custom extension. Some other extension may need 2 lists and a checkbox... no one knows!

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java [hunk 4]
```diff
@@ -30,23 +30,24 @@
...
-                               policy = ReferencePolicy.DYNAMIC,
-                               cardinality = ReferenceCardinality.OPTIONAL_MULTIPLE),
-
-                        @Reference(name = HttpCacheEngineImpl.METHOD_NAME_TO_BIND_CACHE_STORE,
-                               referenceInterface = HttpCacheStore.class,
-                               policy = ReferencePolicy.DYNAMIC,
-                               cardinality = ReferenceCardinality.MANDATORY_MULTIPLE)})
-
+@References({
+        @Reference(name = HttpCacheEngineImpl.METHOD_NAME_TO_BIND_CONFIG,
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java:36]: Removed the required config for this service, but made all the references mandatory. 

This means at least 1 of each of the 3 references must exists for this to be satisfied. WDYT?

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java [hunk 5]
```diff
@@ -208,23 +209,28 @@ public boolean isRequestCacheable(SlingHttpServletRequest request) throws HttpCa
         for (HttpCacheConfig cacheConfig : cacheConfigs) {
             if (cacheConfig.accepts(request)) {
                 isRequestCacheable = true;
+                break;
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java:212]: Minor perf improvement

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java [hunk 6]
```diff
@@ -208,23 +209,28 @@ public boolean isRequestCacheable(SlingHttpServletRequest request) throws HttpCa
...
         }
 
-        // Execute custom rules.
-        for (HttpCacheHandlingRule rule : cacheHandlingRules) {
-            if (!rule.onRequestReceive(request)) {
-                log.debug("Request cannot be cached for the uri {} honoring the rule {}", request.getRequestURI(),
-                        rule.getClass().getName());
-                isRequestCacheable = false;
+        // Execute custom rules as long as one cache config accepts the request
+        if (isRequestCacheable) {
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java:217]: Minor perf improvement

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/filter/impl/HttpCacheFilter.java [hunk 7]
```diff
@@ -26,7 +26,7 @@
         description = "Intercepts http requests to deal with caching.",
         generateComponent = true,
         generateService = true,
-        order = 0,
+        order = 4999, // Magic number places it immediately after the Granite inner cache
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/filter/impl/HttpCacheFilter.java:29]: This pushes the filter up fairly high on the filter stack.. its right after the Granite http inner cache filter

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/filter/impl/HttpCacheFilter.java [hunk 8]
```diff
@@ -54,6 +56,10 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
                 if (cacheEngine.isCacheHit(slingRequest, cacheConfig)) {
                     // Deliver the response from cache.
                     if (cacheEngine.deliverCacheContent(slingRequest, slingResponse, cacheConfig)) {
+                        if (log.isDebugEnabled()) {
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/filter/impl/HttpCacheFilter.java:59]: This is just fun to see how fast it is :)

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCacheMBean.java [hunk 9]
```diff
@@ -23,11 +24,12 @@
     @Description("Cache TTL in Seconds. -1 value represent no TTL.")
     long getTtl();
 
+    @Description("Cache entry contents by key.")
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCacheMBean.java:28]: Added a handy Mbean method so we can inspect the contents of the cache.

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCachePersistenceObject.java [hunk 10]
```diff
@@ -88,7 +91,15 @@ public String getContentType() {
      * @return Returned in <code>Map<String, List<String>></code> format.
      */
     public Map<String, List<String>> getHeaders() {
-        return (Map<String, List<String>>) (Map<?, ?>) Multimaps.asMap(headers);
+        Map<String, List<String>> map = new HashMap<>();
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCachePersistenceObject.java:94]: Ran into exceptions; @Sivaramvt  @jasiwalgaurav can you take a look at this and make sure i didnt much anything up?

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java [hunk 11]
```diff
@@ -28,10 +49,14 @@
...
-                       value = HttpCacheStore.VALUE_MEM_CACHE_STORE_TYPE,
-                       propertyPrivate = true), @Property(name = "jmx.objectname",
-                                                          value = "com.adobe.acs.httpcache:type=CRX")})
+@Properties({
+        @Property(name = HttpCacheStore.KEY_CACHE_STORE_TYPE,
+                    value = HttpCacheStore.VALUE_MEM_CACHE_STORE_TYPE,
+                    propertyPrivate = true),
+        @Property(name = "jmx.objectname",
+                    value = "com.adobe.acs.httpcache:type=In Memory",
+                    propertyPrivate = true)
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java:58]: set to private else was showing in the OSGi configMgr as editable

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java [hunk 12]
```diff
@@ -160,9 +185,15 @@ public void invalidate(HttpCacheConfig cacheConfig) {
...
-                // If matches, invalidate that particular key.
-                cache.invalidate(key);
+            try {
+                if (cacheConfig.knows(key)) {
+                    // If matches, invalidate that particular key.
+                    cache.invalidate(key);
+                }
+            } catch (HttpCacheKeyCreationException e) {
+                // TODO Verify full invalidation is the best approach in the event of a failure
+                log.error("Could not invalidate cache. Falling back to full cache invalidation.", e);
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java:195]: @jasiwalgaurav @Sivaramvt WDYT about this? This exception wasnt being handled so I did this .. im cool w whatever we do w this though; invalidating everything seemed like the safest bet.

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java [hunk 13]
```diff
@@ -201,12 +232,35 @@ public long getTtl() {
     }
 
     @Override
+    public String getCacheEntry(String cacheKeyStr) throws IOException {
+        CacheKey cacheKey = null;
+
+        for (CacheKey cacheKeyTmp : cache.asMap().keySet()) {
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java:238]: Couldnt find a clean way to collect the write params to create a synthetic CacheKey obj. Figured this should work 98% of the time .. ?

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java [hunk 14]
```diff
@@ -214,7 +268,7 @@ public TabularData getCacheStats() throws OpenDataException {
                 .LONG});
 
         final TabularDataSupport tabularData = new TabularDataSupport(new TabularType("cacheEntries", "Cache " +
-                "Entries", cacheEntryType, new String[]{}));
+                "Entries", cacheEntryType, new String[]{"averageLoadPenalty"}));
```
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java]: The last param cannot be empty. I think we want to flip this TabularData eventually. IIRC this last String[] are the set of cols that define unique row in the table (which in this case would be all cols?). Can revist later; wanted to get the MBean displaying atleast.

**Comment by davidjgonzalez** [comment]
@Sivaramvt @jasiwalgaurav truly awesome stuff :)

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/GroupHttpCacheConfigExtension.java (hunk 1 above)
Comment (Sivaramvt) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/GroupHttpCacheConfigExtension.java:42]: I made this config factory and move along with cacheconfig (n - n mapping). Result is so beautiful. For the same uri, i was able to cache separately for 2 different groups.

cache-config-geometrix
cache-config-admin

sling:OSGi config nodes for GroupHttpCacheConfigExtension
com.adobe.acs.commons.httpcache.config.impl.GroupHttpCacheConfigExtension-admins.xml
<?xml version="1.0" encoding="UTF-8"?>
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
          jcr:primaryType="sling:OsgiConfig"
          config.name="group-admins"
          httpcache.config.extension.user-groups="administrators"/>

com.adobe.acs.commons.httpcache.config.impl.GroupHttpCacheConfigExtension-geo.xml
<?xml version="1.0" encoding="UTF-8"?>
<jcr:root xmlns:sling="http://sling.apache.org/jcr/sling/1.0" xmlns:jcr="http://www.jcp.org/jcr/1.0"
          jcr:primaryType="sling:OsgiConfig"
          config.name="group-geo"
          httpcache.config.extension.user-groups="geometrixx-twitter"/>

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/GroupHttpCacheConfigExtension.java (hunk 2 above)
Comment (Sivaramvt) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/GroupHttpCacheConfigExtension.java:88]: +1. Now am thoroughly satisfied. When I make this GroupHttpCacheConfigExtension a config factory, am able to cache same URI for 2 different groups with  2 cache configs. :)

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java (hunk 3 above)
Comment (Sivaramvt) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/config/impl/HttpCacheConfigImpl.java:114]: Agreed.

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java (hunk 4 above)
Comment (Sivaramvt) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java:36]: That boils to down to mandating atleast one cache config to be present to get this active. +1.

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCacheMBean.java (hunk 9 above)
Comment (Sivaramvt) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCacheMBean.java:28]: JMX beans working. So excited to see my cache keys and content.

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCachePersistenceObject.java (hunk 10 above)
Comment (Sivaramvt) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCachePersistenceObject.java:94]: Notes. Added a FIXME in the code. Will check.

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java (hunk 12 above)
Comment (Sivaramvt) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java:195]: Yes. It's the safest option.

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java (hunk 13 above)
Comment (Sivaramvt) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemHttpCacheStoreImpl.java:238]: I think this is probably the only easy way.

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCachePersistenceObject.java (hunk 10 above)
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/store/mem/impl/MemCachePersistenceObject.java:94]: Not sure if we need same keys; maybe this should be a multi map that has a List<String> and not a hashmap?

File: bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java (hunk 4 above)
Comment (davidjgonzalez) [inline bundle-httpcache/src/main/java/com/adobe/acs/commons/httpcache/engine/impl/HttpCacheEngineImpl.java:36]: Should cache handling rules be optional? Config and persistence should be mandatory.

**Pull request merged**